frame/election-providers/src/onchain.rs

```diff
@@ -0,0 +1,190 @@
+use sp_election_providers::{ElectionDataProvider, ElectionProvider};
+use sp_npos_elections::{
+	ElectionResult, ExtendedBalance, IdentifierT, PerThing128, Supports, VoteWeight,
+};
+use sp_runtime::RuntimeDebug;
+use sp_std::{collections::btree_map::BTreeMap, marker::PhantomData, prelude::*};
+
+/// Errors of the on-chain election.
+#[derive(RuntimeDebug, Eq, PartialEq)]
+pub enum Error {
+	/// An internal error in the NPoS elections crate.
+	NposElections(sp_npos_elections::Error),
+	/// An assignment failed to pass the feasibility check
+	Feasibility(&'static str),
+}
+
+impl From<sp_npos_elections::Error> for Error {
+	fn from(e: sp_npos_elections::Error) -> Self {
+		Error::NposElections(e)
+	}
+}
+
+/// A simple on-chian implementation of the election provider trait.
+///
+/// This will accept voting data on the fly and produce the results immediately.
+///
+/// ### Warning
+///
+/// This can be very expensive to run frequently on-chain. Use with care.
+pub struct OnChainSequentialPhragmen<T: Config>(PhantomData<T>);
+
+/// Configuration trait of [`OnChainSequentialPhragmen`].
+///
+/// Note that this is similar to a pallet traits, but [`OnChainSequentialPhragmen`] is not a pallet.
+pub trait Config {
+	/// The account identifier type.
+	type AccountId: IdentifierT;
+	/// The block number type.
+	type BlockNumber;
+	/// The accuracy used to compute the election:
+	type Accuracy: PerThing128;
+	/// Something that provides the data for election.
+	type DataProvider: ElectionDataProvider<Self::AccountId, Self::BlockNumber>;
+}
+
+impl<T: Config> ElectionProvider<T::AccountId, T::BlockNumber> for OnChainSequentialPhragmen<T>
+where
+	ExtendedBalance: From<<T::Accuracy as PerThing>::Inner>,
+{
+	type Error = Error;
+	type DataProvider = T::DataProvider;
+
+	fn elect() -> Result<Supports<T::AccountId>, Self::Error> {
+		let voters = Self::DataProvider::voters();
+		let targets = Self::DataProvider::targets();
+		let desired_targets = Self::DataProvider::desired_targets() as usize;
+
+		let mut stake_map: BTreeMap<T::AccountId, VoteWeight> = BTreeMap::new();
+
+		voters.iter().for_each(|(v, s, _)| {
+			stake_map.insert(v.clone(), *s);
+		});
+
+		let stake_of = Box::new(|w: &T::AccountId| -> VoteWeight {
+			stake_map.get(w).cloned().unwrap_or_default()
+		});
+
+		let ElectionResult {
+			winners,
+			assignments,
+		} = sp_npos_elections::seq_phragmen::<_, T::Accuracy>(desired_targets, targets, voters, None)
+			.map_err(Error::from)?;
+
+		// check all assignments for feasibility, based on election data provider.
+		assignments
+			.iter()
+			.map(Self::DataProvider::feasibility_check_assignment)
+			.collect::<Result<_, _>>()
+			.map_err(|e| Error::Feasibility(e))?;
+
+		let staked =
+			sp_npos_elections::assignment_ratio_to_staked_normalized(assignments, &stake_of)?;
+		let winners = sp_npos_elections::to_without_backing(winners);
+
+		sp_npos_elections::to_supports(&winners, &staked).map_err(Error::from)
+	}
+
+	fn ongoing() -> bool {
```

Comment:
This API is added to cope with deferring slashes during election. I believe now we can do that as election no longer uses live staking data, but rather a snapshot thereof. Although it is harmless to keep around as well and remove it later as it would require some changes in offences pallet as well.